### PR TITLE
Add Oauth2 login to site.

### PR DESCRIPTION
**Pull request by Zwork101**
The feature we've all been waiting for...

### CODE REVIEW

File: pysite/constants.py [hunk 1]
```diff
@@ -20,3 +21,7 @@ class ValidationTypes(Enum):
 ADMIN_ROLE = 267628507062992896
 MODERATOR_ROLE = 267629731250176001
 HELPER_ROLE = 267630620367257601
+
+CLIENT_ID = environ.get('CLIENT_ID', '')
+CLIENT_SECRET = environ.get('CLIENT_SECRET', '')
+SCOPE = 'identify email'
```
Comment (gdude2002) [inline pysite/constants.py:27]: Too generic - can we prefix all of these with, eg, `DISCORD_OAUTH`? Env vars too.

File: pysite/views/error_handlers/http_404.py [hunk 2]
```diff
@@ -9,4 +9,4 @@ class Error404View(ErrorView):
     error_code = 404
 
     def get(self, error: NotFound):
-        return "replace me with a template, 404 not found", 404
+        return "Unable to find page", 404
```
Comment (gdude2002) [inline pysite/views/error_handlers/http_404.py:12]: This change is out of scope for this PR. Also, should be templated. :P

**Review by gdude2002 — CHANGES_REQUESTED** [review]
Huh, this is really all there is? We'll definitely have to expand on this.

File: pysite/views/error_handlers/http_404.py (hunk 2 above)
Comment (Zwork101) [reply pysite/views/error_handlers/http_404.py:12]: Yeah, sorry. I had it changed, but then I needed to revert it back, and I forgot what it originally said.

File: pysite/constants.py (hunk 1 above)
Comment (Zwork101) [reply pysite/constants.py:27]: Alrighty

**Review by gdude2002 — DISMISSED** [review]
Tentatively approving - I'm not 100% sure this is entirely ready for merge to master just yet as it's not functional without several views and database integration, but it won't hurt to add it either.

**Comment by gdude2002** [comment]
Wow, that coverage change. That is.. wow.

**Review by gdude2002 — DISMISSED** [review]
OK, so here's what needs doing before we consider this to be in a proper mergeable state.

* Views: Login, logout, authorized, error
    * My advice is to place these all under the same path, so eg `/auth/login`, `/auth/logout`, etc
* Session persistence: Sessions **must** be persisted in the database. They should have their own table, mapping session keys to the users table via user IDs
* Deny login to users that aren't on our server - there's an OAuth scope for this - and instruct them on how to rectify that problem
* Add login to the navbar, and display the logged-in user in place of the login link (it could be used to logout in that case)

File: pysite/route_manager.py [hunk 3]
```diff
@@ -5,10 +5,14 @@
 import os
 
 from flask import Blueprint, Flask
+from flask_dance.contrib.discord import make_discord_blueprint
 from flask_sockets import Sockets
 
 from pysite.base_route import APIView, BaseView, ErrorView, RouteView
+from pysite.constants import DISCORD_OAUTH_ID, DISCORD_OAUTH_SCOPE, \
```
Comment (gdude2002) [inline pysite/route_manager.py]: Please use parens for line splitting, not backslahes.

```py
from x import (
    a, b, c, d
)
```

File: pysite/oauth.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+from flask_dance.contrib.discord import discord
+import requests
+
+from pysite.constants import DISCORD_API_ENDPOINT, SERVER_ID
+from pysite.mixins import DBMixin
+
+_MAIN_BACK = None  # type: pysite.database.RethinkDB
+
+
+class OauthBackend(BaseBackend, DBMixin):
```
Comment (gdude2002) [inline pysite/oauth.py]: I don't understand what this class does - can you comment it?

**Comment by gdude2002** [comment]
Missing a feature:

> * Add login to the navbar, and display the logged-in user in place of the login link (it could be used to logout in that case)

I don't fully understand how session persistence is working, here - can you walk me through it?

**Review by gdude2002 — CHANGES_REQUESTED** [review]
See comments

File: app_test.py [hunk 5]
```diff
@@ -79,6 +80,21 @@ def test_ws_test(self):
         response = self.client.get('/ws_test')
         self.assertEqual(response.status_code, 200)
 
+    def test_oauth_login(self):
+        """check oauth redirects """
```
Comment (jerbob) [inline app_test.py]: These Docstrings aren't very consistent. I think we'll need something like:
```py
""" Check OAuth redirects """
```

File: app_test.py [hunk 6]
```diff
@@ -79,6 +80,21 @@ def test_ws_test(self):
...
+        response = self.client.get(DISCORD_OAUTH_REDIRECT)
+        self.assertEqual(response.status_code, 302)
+
+    def test_oauth_logout(self):
+        """check oauth redirects """
```
Comment (jerbob) [inline app_test.py]: Can we have docstrings like this:
```py
""" Check OAuth redirects """
```

File: app_test.py [hunk 7]
```diff
@@ -79,6 +80,21 @@ def test_ws_test(self):
...
+        response = self.client.get('/auth/logout')
+        self.assertEqual(response.status_code, 302)
+
+    def test_oauth_authorized(self):
+        """check oauth authorization"""
```
Comment (jerbob) [inline app_test.py]: Docstring consistency again :D 
```py
""" Check OAuth authorization"""
```

File: app_test.py [hunk 8]
```diff
@@ -29,7 +30,7 @@ def create_app(self):
 class BaseEndpoints(SiteTest):
     """ test cases for the base endpoints """
```
Comment (jerbob) [inline app_test.py]: Docstring consistency pls
```py
""" Test cases for the base endpoints """
```

File: pysite/constants.py [hunk 9]
```diff
@@ -21,6 +22,17 @@ class ValidationTypes(Enum):
...
+
+DISCORD_API_ENDPOINT = "https://discordapp.com/api"
+
+DISCORD_OAUTH_REDIRECT = "/auth/discord"
+DISCORD_OAUTH_AUTHORIZED = "/auth/discord/authorized"
+DISCORD_OAUTH_ID = environ.get('DISCORD_OAUTH_ID', '')
+DISCORD_OAUTH_SECRET = environ.get('DISCORD_OAUTH_SECRET', '')
+DISCORD_OAUTH_SCOPE = 'identify email guilds.join'
+
+
```
Comment (jerbob) [inline pysite/constants.py]: Any reason why we have this extra space?

File: pysite/oauth.py [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+    def get(self, *args, **kwargs):  # Not used
+        pass
+
+    def set(self, blueprint, token):
+        user = get_user()  # get the user's information, based on a local proxy
+        join_discord(token["access_token"], user["id"])  # Attempt to join a discord
+        sess_id = str(uuid5(uuid4(), self.key))  # Create a session ID, encrypted to protect against timing attacks
+        session["session_id"] = sess_id  # Set it
+        self.db.insert("oauth_data", {"id": sess_id, "access_token": token["access_token"], "snowflake": user['id'],
+                                      "refresh_token": token["refresh_token"], "expires_at": token["expires_at"]})
```
Comment (jerbob) [inline pysite/oauth.py]: Can we have something more readable here and for the insert below? Something like:
```py
self.db.insert("oauth_data", {
        "id": sess_id,
        "access_token": token["access_token"],
        "snowflake": user['id'],
        "refresh_token": token["refresh_token"],
        "expires_at": token["expires_at"]
    }
)
```

File: pysite/oauth.py [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+        creds = _MAIN_BACK.db.get("oauth_data", id)
+        if creds:
+            return _MAIN_BACK.db.get("users", creds["snowflake"])
+    elif not _MAIN_BACK:
+        logging.warning("Failed to get user data: _MAIN_BACK not loaded")
+
+
+def logout():
+    sess_id = session.get("session_id")
+    if sess_id and _MAIN_BACK.db.get("oauth_data", sess_id):  # If user exists in db,
```
Comment (jerbob) [inline pysite/oauth.py]: Can this inline comment be above the line? It might be worth making the comments more descriptive too (aka `remove the user's session` rather than `remove them (at least, their session)`)

**Review by jerbob — CHANGES_REQUESTED** [review]

File: pysite/database.py [hunk 12]
```diff
@@ -103,6 +103,34 @@ def create_table(self, table_name: str, primary_key: str="id", durability: str="
             self.log.debug(f"Table created: '{table_name}'")
             return True
 
+    def delete(self, table_name: str, primary_key: Union[str, None] = None,
+               durability: str="hard", return_changes: Union[bool, str] = False
+               ) -> Union[Dict[str, Any], None]:
```
Comment (jerbob) [inline pysite/database.py]: Can we change this to something like:
```py
def delete(
    self,
    table_name: str,
    primary_key: Union[str, None] = None,
    durability: str="hard",
    return_changes: Union[bool, str] = False    
) -> Union[Dict[str, Any], None]:
```

**Review by jerbob — CHANGES_REQUESTED** [review]

File: pysite/constants.py (hunk 1 above)
Comment (gdude2002) [reply pysite/constants.py:27]: Did you get around to looking into this?

File: pysite/views/error_handlers/http_404.py (hunk 2 above)
Comment (gdude2002) [reply pysite/views/error_handlers/http_404.py:12]: Won't this conflict with joph's stuff now?

**Review by gdude2002 — CHANGES_REQUESTED** [review]
Where's the database stuff gone?
